### PR TITLE
Fix presentation of HKT and TypeLambdas in Scala3doc

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

**Review by abgruszecki — COMMENTED** [review]
This seems to improve some things, so I think we can merge it on that basis. However, we still display type lambdas instead of HKTs, see all the `exnteds` clauses here: https://scala3doc.virtuslab.com/pr-missing-wildcard-in-hkts/scala3/api/scala/collection/immutable.html. That should be considered a bug, and isn't fixed here.

File: scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 1]
```diff
@@ -409,7 +409,7 @@ trait ClassLikeSupport:
       if  argument.symbol.flags.is(Flags.Covariant) then "+"
       else if argument.symbol.flags.is(Flags.Contravariant) then "-"
       else ""
-    val name = argument.symbol.normalizedName
+    val name = argument.symbol.normalizedName.takeWhile(_ != '$')
```
Comment (romanowski) [inline scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala:412]: One can use `$` in type name, eg. `class A[A$B]` compiles fine.

File: scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 2]
```diff
@@ -418,15 +418,17 @@ trait ClassLikeSupport:
...
 
-  def parseTypeDef(typeDef: TypeDef): Member =
+  def parseTypeDef(c: ClassDef, typeDef: TypeDef): Member =
     def isTreeAbstract(typ: Tree): Boolean = typ match {
       case TypeBoundsTree(_, _) => true
       case LambdaTypeTree(params, body) => isTreeAbstract(body)
       case _ => false
     }
 
+    val memberInfo = unwrapMemberInfo(c, typeDef.symbol)
```
Comment (romanowski) [inline scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala:428]: Why do we need `memberInfo`? It does not seems to be used anywhere in code.

File: scala3doc/src/dotty/dokka/tasty/TypesSupport.scala [hunk 3]
```diff
@@ -102,9 +102,10 @@ trait TypesSupport:
       case AnnotatedType(tpe, _) =>
         inner(tpe)
       case tl @ TypeLambda(params, paramBounds, resType) =>
-        // println(params)
-        // println(paramBounds)
-        texts("[") ++ commas(params.zip(paramBounds).map( (name, typ) => texts(s"${name}") ++ inner(typ) )) ++ texts("]")
+        texts("[") ++ commas(params.zip(paramBounds).map { (name, typ) =>
+          val normalizedName = name.takeWhile(_ != '$')
```
Comment (romanowski) [inline scala3doc/src/dotty/dokka/tasty/TypesSupport.scala:106]: `$` can be part of appliedType, `class A[+CC <: ([A$B] =>> Any)]` compiles

**Review by romanowski — CHANGES_REQUESTED** [review]
We cannot simply strip anything after `$` in type names

File: scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 1 above)
Comment (BarkingBad) [reply scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala:412]: Aleksander told me that using `$` sign in names in Scala is widely unrecommended and no one guarantees that everything works fine, though I can chagne it to be the reqex matching `_\$\d*`

**Comment by BarkingBad** [comment]
@abgruszecki I am not sure if it is correct behaviour, but look at this example, given code like this:
```scala
import scala.collection.immutable.ArraySeq
trait SomeTraitWithHKTs[A[_], B[_], C[_], D[_], E[_]]
abstract class Foo[Z]
trait SomeTrait extends SomeTraitWithHKTs[Foo, scala.collection.immutable.ArraySeq, ArraySeq, List, [B] =>> List[B]]
```

we get following `AppliedType`'s parameter list:
<details>
  <summary>Full output</summary>
  
`
List(HKTypeLambda(List(Z), List(TypeBounds(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Nothing),TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Any))), AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class tests)),object hkts),Foo),List(TypeParamRef(Z)))), TypeRef(TermRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),object collection),object immutable),class ArraySeq), HKTypeLambda(List(A), List(TypeBounds(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Nothing),TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Any))), AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class collection)),object immutable),ArraySeq),List(TypeParamRef(A)))), TypeRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),package),List), HKTypeLambda(List(B), List(TypeBounds(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Nothing),TypeRef(TermRe
`
</details>

which is actually List of `[HKTypeLambda, TypeRef, HKTypeLambda, TypeRef, HKTypeLambda]` and the output is:
obraz

1. There is different behaviour to `ArraySeq` declared as an import or as fully qualified name making it just `ArraySeq` or type lambda of `[A] =>> ArraySeq[A]`.
2. If i would like to collapse the `Foo` type lamda, I could not disambiguate it with actual type lambdas, like ` [B] =>> List[B]`. 

What do you think about these two things?

File: scala3doc/src/dotty/dokka/tasty/TypesSupport.scala (hunk 3 above)
Comment (BarkingBad) [reply scala3doc/src/dotty/dokka/tasty/TypesSupport.scala:106]: Changed to regex

File: scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 2 above)
Comment (BarkingBad) [reply scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala:428]: My bad, fixed

File: scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 1 above)
Comment (BarkingBad) [reply scala3doc/src/dotty/dokka/tasty/ClassLikeSupport.scala:412]: Changed to regex

**Comment by abgruszecki** [comment]
Ah, I see, in some cases we do display `List` as simply `List`. Do we have what you wrote as a test case? If not, we should.

At some point we should find out what exactly makes the types in stdlib display type lambdas instead of HKTs (variance annotations, perhaps?), but it's less of a problem than I thought, I'd say.

**Comment by BarkingBad** [comment]
I am still not sure why this thing is happening, but I have shown the example where import of fullly qualified name makes difference for `ArraySeq`. 
@romanowski the memberInfo was the unfinished work. However, there come out another weird problem, for given code:
```scala
trait A[X] {
  type SomeType[X]
  def x: SomeType[X]
}

trait B[D] extends A[D]
```

the expected behaviour to see `SomeType` from `B` class perspective would be with applied `D` symbolic name, not `X`. However, we get following, because `memberInfo` method returns such name though it is run in context of `B` class symbol.
obraz

**Comment by BarkingBad** [comment]
Unfortunately I am enforced to rollback all `memberInfo` changes for `TypeDefs` as it returns `HKTypeLambdas` which lose variance

**Pull request merged**